extgrpc/ext_grpc.go

```diff
@@ -98,7 +123,66 @@ func decodeWithGrpcCode(
 	return &withGrpcCode{cause: cause, code: codes.Code(wp.Code)}
 }
 
+// encodeGrpcStatus takes an error generated by a standard gRPC Status and
+// converts it into a GoGo Protobuf representation from
+// github.com/gogo/googleapis/google/rpc.
+//
+// This is necessary since EncodedError uses an Any field for structured errors,
+// and thus can only contain Protobufs that have been registered with the GoGo
+// Protobuf type registry -- the standard gRPC Status type is not a GoGo
+// Protobuf, and is therefore not registered with it and cannot be decoded by
+// DecodeError().
+//
+// Also note that in order to use error details, the input type must be a
+// gogoproto Status from github.com/gogo/status, not from the standard gRPC
+// Status, and all details must be gogoproto types. The reasons for this
+// are the same as for the Any field issue mentioned above.
+func encodeGrpcStatus(_ context.Context, err error) (string, []string, proto.Message) {
+	s := gogostatus.Convert(err)
+	// If there are known safe details, return them.
+	details := []string{}
+	for _, detail := range s.Details() {
+		if safe, ok := detail.(errbase.SafeDetailer); ok {
+			details = append(details, safe.SafeDetails()...)
+		}
+	}
+	return s.Message(), details, s.Proto()
+}
+
+// decodeGrpcStatus is the inverse of encodeGrpcStatus. It takes a gogoproto
+// Status as input, and converts it into a standard gRPC Status error.
+func decodeGrpcStatus(
+	ctx context.Context, msg string, details []string, payload proto.Message,
+) error {
+	return grpcstatus.Convert(decodeGoGoStatus(ctx, msg, details, payload)).Err()
```

Comment:
I believe this is incorrect (or maybe I am confusing myself): we need to distinguish the case where the error decode fails, from the decoded error.

IIRC the API states that if the error decoding fails, the decoding function should return nil. Right now you wrote it to return an unrelated error.

Reply:
Good point, changed `decodeGoGoStatus` to return `nil` if the input type was unexpected. This generally should never happen anyway, since it's specified in `RegisterLeafEncoder`.

Reply:
unless someone in the future makes a mistakes and migrates the package to a new name without registering the migration, or something else. These things here are also guardrails for future changes.